kernel/src/engine/sync/fs_client.rs

```diff
@@ -51,19 +51,19 @@ impl FileSystemClient for SyncFilesystemClient {
                 .sorted_by_key(|ent| ent.path())
                 .map(|ent| {
                     ent.metadata().map_err(Error::IOError).and_then(|metadata| {
-                        let last_modified: u64 = metadata
+                        let last_modified = metadata
                             .modified()
-                            .map(
-                                |modified| match modified.duration_since(SystemTime::UNIX_EPOCH) {
-                                    Ok(d) => d.as_secs(),
-                                    Err(_) => 0,
-                                },
-                            )
+                            .ok()
+                            .and_then(|modified| {
+                                modified.duration_since(SystemTime::UNIX_EPOCH).ok()
+                            })
+                            .and_then(|modified| modified.as_millis().try_into().ok())
                             .unwrap_or(0);
```

Comment:
Are we converting Result to Option three times here? Why is that needed? 
```suggestion
                            .and_then(|modified| modified.duration_since(SystemTime::UNIX_EPOCH))
                            .and_then(|modified| modified.as_millis().try_into())
                            .unwrap_or(0);
```
Or do types not match in some annoying way?

Reply:
Honestly, all this monadic chaining feels like a poor substitute for a missing helper method...

```rust
impl TryFrom<DirEntry> for FileMeta {
    type Error = Error;

    fn try_from(ent: DirEntry) -> DeltaResult<FileMeta> {
        let metadata = ent.metadata()?;
        let last_modified = metadata.modified()?.duration_since(SystemTime::UNIX_EPOCH)?;
        Ok(FileMeta {
            location: Url::from_file_path(ent.path())?,
            last_modified: last_modified.as_millis().try_into()?,
            size: metadata.len() as usize,
        })
    }
}
```
and then
```rust

            let it = all_ents
                .into_iter()
                .sorted_by_key(|ent| ent.path())
                .map(TryFrom::try_from);
```

Reply:
Woah that's clean 👌  I'm doing a lot of map_err, and I'm wondering if we should do `From` conversions for common things like `Url::from_*` and `TryInto` 🤔

Reply:
yea seems reasonable? make an issue for us to think about?

Reply:
We already _have_ the `From` conversion for `url::ParseError`, so `?` should Just Work? The problem with `TryInto` is the error type is parametrized, so there's no way to know in advance what might be needed. We'd have to go case by case (when in doubt, try `?` and hopefully there's already an `impl From` for it to use).